### ISSUE TITLE
Using the GitLab and GitHub API to pull in metrics about code changes

### ISSUE THREAD

**Issue by hsu000001** [usgs-biolab/pymodelcat#18]
The CIG (Computational Infrastructure for Geodynamics) software pages pull in "Code changes:
X commits this past month, X commits this past year" to their individual software pages (example: https://geodynamics.org/cig/software/aspect/). I asked about this and was told it was pulled in from the GitHub API. 

A requested feature for the USGS Model Catalog was to pull in metrics such as these to see how "active" the repos are. What metrics are useful and how should they be labeled or described to paint an accurate picture about "community use" "activity" and so on?

Gitlab API: https://docs.gitlab.com/ee/api/README.html
GitHub API: https://docs.github.com/en/rest

I would need help from someone to decode what metrics are available, and how to describe them.

Are these useful and available on both platforms? Are month and year the appropriate time frames? What is missing?
X commits past month
X commits past year
X issues opened (past month, past year)
X issues commented on (past month, past year)
X pull requests (past month, past year)

**Comment by skybristol** [comment]
Yes, the GitHUB REST api is nicely navigable and provides a ton of information. If you've never looked at it, check out this great [GitHub visualizer tool](http://ghv.artzub.com/#repo=pymodelcat&climit=100&user=usgs-biolab). Or if you want to really blow your mind, make a cocktail, get comfortable, and run the full history for [Facebook's React](http://ghv.artzub.com/#repo=react&climit=100&user=facebook). That's all driven by GitHub's API and mostly the events route.

There are probably a whole raft of metrics that could be useful to analyze and assess the relative health, maturity, and activity of a given model development effort. The metrics you suggest seem like a reasonable start. I would also consider the following as interesting initial stuff:

- [Forks](https://api.github.com/repos/usgs-biolab/pymodelcat/forks) (I'd pull something like user, url to fork, and date(s); it's a somewhat interesting gage of some level of active interest)
- [Events](https://api.github.com/repos/usgs-biolab/pymodelcat/events) could show a basic timeline of events playing out for a repository that would include not just code commits but issue actions by specific users over time
- Watchers and Stargazers - could just report the number of, but the API will show details on who has starred or is watching the repo

Here's a [paper](https://www.frontiersin.org/articles/10.3389/fbioe.2018.00198/full) from a couple years ago that explores some of this in a biomed context.

Another question is how decisions here should be implemented. We could provide a REST API abstraction through pymodelcat code as a lightweight wrapper on the REST API(s) that filters a little bit and provides some basic business logic for what metrics mean the most for our context. That would also provide some level of harmonization across GitHub and GitLab instances, though not everything is going to be available equally across the two different platforms. We could go so far as to cache the metrics we care about somewhere, but I would avoid that if at all possible and just do everything dynamically. 

It might be cool to eventually come up with some kind of activity index or health index, maybe something multi-factor and compounding, that would result in a ranking system along different tracks. That could be a cool ESIPLab project some day, eh?

**Comment by trashbirdecology** [comment]
Might also be wise to allow the developers to throw a lifecycle badge up on
the readme, which I am guessing you could mind from the repository files?

>
